src/gnc_attitude_estimator.cpp

```diff
@@ -28,16 +28,348 @@
+    ukf_float weight_o = 1.0f / (2.0f * (N + lambda));
+
+    // Calculate x_bar and z_bar
+    x_bar = weight_c * state.sigmas[0];
+    z_bar = weight_c * state.measures[0];
+    for (lin::size_t i = 1; i < 13; i++) {
+      x_bar = x_bar + weight_o * state.sigmas[i];
+      z_bar = z_bar + weight_o * state.measures[i];
+    }
+
+    // Plus the associated covariances
+    UkfVector6 dx1 = state.sigmas[0] - x_bar;
+    UkfVector5 dz1 = state.measures[0] - z_bar;
+    P_bar = P_bar + weight_c * dx1 * lin::transpose(dx1);
+    P_vv = P_vv + weight_c * dz1 * lin::transpose(dz1);
+    P_xy = P_xy + weight_c * dx1 * lin::transpose(dz1);
+    for (lin::size_t i = 1; i < 13; i++) {
+      dx1 = state.sigmas[i] - x_bar;
+      dz1 = state.measures[i] - z_bar;
+      P_bar = P_bar + weight_o * (dx1 * lin::transpose(dx1));
+      P_vv = P_vv + weight_o * (dz1 * lin::transpose(dz1));
+      P_xy = P_xy + weight_o * (dx1 * lin::transpose(dz1));
+    }
+
+    // Sensor noise covariance
+    UkfMatrix5x5 R = lin::zeros<UkfMatrix5x5>();
+    {
+      R(0, 0) = constant::ukf_sigma_s * constant::ukf_sigma_s;
+      R(1, 1) = R(0, 0);
+      R(2, 2) = constant::ukf_sigma_b * constant::ukf_sigma_b;
+      R(3, 3) = R(2, 2);
+      R(4, 4) = R(2, 2);
+    }
+
+    P_bar = P_bar + Q; // Add process noise to predicted covariance
+    P_vv = P_vv + R;   // Add sensor noise to innovation covariance
+  }
+
+  // Kalman gain step
+  ukf_kalman_update(state, data);
+
+  // Process x (which is now x_new) and P (which is now P_new)
+  {
+    // Update time
+    state.t = data.t;
+
+    // Perturb q_new according to the new state
+    lin::Vector4d q;
+    utl::grp_to_quat(lin::ref<3, 1>(state.x, 0, 0).eval(), a, f, q);
+    utl::quat_cross_mult(q, q_new, state.q);
+
+    // Reset the attitude portion of the state to zeros
+    lin::ref<3, 1>(state.x, 0, 0) = lin::zeros<UkfVector3>();
+  }
+}
+
+/** @fn ukf_m
+ *  Update attitude estimator state given a magnetometer reading.
+ * 
+ *  @param[inout] state Attitude filter state.
+ *  @param[in]    data  Input sensor data. */
+static void ukf_m(AttitudeEstimatorState &state, AttitudeEstimatorData const &data) {
+  // TODO : Implement this
+  state = AttitudeEstimatorState();
+}
+
+/** @fn ukf_ms
+ *  Update attitude estimator state given magnetometer and sun vector readings.
+ * 
+ *  @param[inout] state Attitude filter state.
+ *  @param[in]    data  Input sensor data. */
+static void ukf_ms(AttitudeEstimatorState &state, AttitudeEstimatorData const &data) {
+  ukf(state, data, [](AttitudeEstimatorState &state, AttitudeEstimatorData const &data) -> void {
+    // Calculate Kalman gain
+    UkfMatrix6x5 K;
+    {
+      UkfMatrix5x5 Q, R;
+      lin::qr(state.P_vv, Q, R);
+      lin::backward_sub(R, Q, lin::transpose(Q).eval()); // Q = inv(P_yy)    
+      K = state.P_xy * Q;
+    }
+
+    // Calculate this steps measurement
+    UkfVector5 z_new {
+      lin::atan(data.s_body(1) / data.s_body(0)),
```

Comment:
This can be NAN if `s_body(0)==0`

Reply:
This is a little concerning... However, in cxx a finite float over zero results in plus or minus infinity (IEEE floats at least) and according to the documentation the `std::atan` function (which is what `lin::atan` essentially calls) returns `pi/2` and `-pi/2` for `inf` and `-inf` respectively. In theory we should be alright.
This code will get replaced when the new sun vector variance handling scheme is incorporated (issue for that coming soon).

Links:
https://en.cppreference.com/w/cpp/numeric/math/atan
https://stackoverflow.com/questions/4745311/c-division-by-0